### ISSUE TITLE
Program Release build fine,Debug build crash

### ISSUE THREAD

**Issue by hongnod** [copperspice/copperspice#251]
I build my program with cs -1.8.1- release -and - debug lib , release build works fine, but debug build crashes with following message 
![image](https://user-images.githubusercontent.com/6680267/217410393-8b481a3f-3dcd-4c31-90f6-2cdb0402ace8.png)

**Comment by hongnod** [comment]
I did some search, it possibly relates to "try_lock" in  "copperspice/src/annex/cs_libguarded/cs_shared_guarded.h"  .

**Comment by hongnod** [comment]
from https://github.com/microsoft/STL/blob/main/stl/src/mutex.cpp
![image](https://user-images.githubusercontent.com/6680267/217469594-1054314d-986a-4893-bdd5-a52749f360e9.png)

**Comment by bgeller** [comment]
When submitting an issue please let us know the following information:

1   What OS are you using..

2   Which compiler are you using and what is the release version.  
( For example,  gcc 10.2, clang 9, or MinGW 10, etc) 

3   If you are using MSVC, which version.  If you are using MinGW on Windows are you using MSys2. 

4   Did you build your application with our CS binary files or did you build CopperSpice from source.

5  When using a debug version of CS did you compile your source in debug mode. 

6   Can you write a small sample which will repeat the issue.

**Comment by bgeller** [comment]
> I did some search, it possibly relates to "try_lock" in "copperspice/src/annex/cs_libguarded/cs_shared_guarded.h" .

The CsLibGuarded library is not new and has been in CS for over six years. This particular header file has been present and unchanged in the last two releases of CopperSpice.  We simply moved this file from CsCore to a new location.

**Comment by bgeller** [comment]
> from https://github.com/microsoft/STL/blob/main/stl/src/mutex.cpp

If this were the issue than it would imply their standard library has a problem. This is unlikely so let's look at how you built CS and/or your application.

**Comment by hongnod** [comment]
> When submitting an issue please let us know the following information:
> 
> 1 What OS are you using..
> Windows10 & Windows11
> 2 Which compiler are you using and what is the release version. ( For example, gcc 10.2, clang 9, or MinGW 10, etc)
> MSVC
> 3 If you are using MSVC, which version. If you are using MinGW on Windows are you using MSys2.
> MSVC 2022.
> 4 Did you build your application with our CS binary files or did you build CopperSpice from source.
> Tested with official binary and custom build from source.
> 5 When using a debug version of CS did you compile your source in debug mode.
> Yes.
> 6 Can you write a small sample which will repeat the issue.
My program is a little bit big and complex, currently can't  figure how to .

**Comment by hongnod** [comment]
> 6 Can you write a small sample which will repeat the issue.

I made it

**Comment by hongnod** [comment]
I verified this issue with more testing.

**Comment by hongnod** [comment]
Wondering if this relative?
https://forum.copperspice.com/viewtopic.php?f=12&t=2637

**Comment by hongnod** [comment]
I did a commit search, found following commit, but I don't know if is it relative.

```
SHA-1: 4b3cc715f450898dd0cbc18eb05ba1af1c6596d1

* Move cs_libguarded source from src/core/guarded to src/annex/cs_libguarded
10c10
< * CopperSpice is distributed in the hope that it will be useful,
---
> * CsLibGuarded is distributed in the hope that it will be useful,
233,234c233,234
<       void rcu_write_lock(const rcu_list<T, M, Alloc> &list);
<       void rcu_write_unlock(const rcu_list<T, M, Alloc> &list);
---
>       void rcu_write_lock(rcu_list<T, M, Alloc> &list);
>       void rcu_write_unlock(rcu_list<T, M, Alloc> &list);
306c306
< void rcu_list<T, M, Alloc>::rcu_guard::rcu_write_lock(const rcu_list<T, M, Alloc> &list)
---
> void rcu_list<T, M, Alloc>::rcu_guard::rcu_write_lock(rcu_list<T, M, Alloc> &list)
308a309
>    list.m_write_mutex.lock();
312c313
< void rcu_list<T, M, Alloc>::rcu_guard::rcu_write_unlock(const rcu_list<T, M, Alloc> &list)
---
> void rcu_list<T, M, Alloc>::rcu_guard::rcu_write_unlock(rcu_list<T, M, Alloc> &list)
313a315
>    list.m_write_mutex.unlock();
555a558,599
> template <typename... Us>
> auto rcu_list<T, M, Alloc>::emplace(const_iterator iter, Us &&...vs) -> iterator
> {
>    auto newNode = detail::allocate_unique<node>(m_node_alloc, std::forward<Us>(vs)...);
> 
>    node *oldHead = m_head.load();
>    node *oldTail = m_tail.load();
> 
>    if (oldHead == nullptr) {
>       // inserting into an empty list
>       m_head.store(newNode.get());
>       m_tail.store(newNode.get());
> 
>    } else if (oldHead == iter.m_current) {
>       // inserting at the beginning of a non-empty list
>       newNode->next.store(oldHead);
>       oldHead->back.store(newNode.get());
>       m_head.store(newNode.get());
> 
>    } else if (oldTail == iter.m_current) {
>       // inserting at the end of a non-empty list
>       newNode->back.store(oldTail);
>       oldTail->next.store(newNode.get());
>       m_tail.store(newNode.get());
> 
>    } else {
>       // inserting in the middle of a non-empty list
>       node *oldBack = iter.m_current->back.load();
> 
>       newNode->next.store(iter.m_current);
>       newNode->back.store(oldBack);
>       iter.m_current->back.store(newNode.get());
> 
>       if (oldBack != nullptr) {
>          oldBack->next.store(newNode.get());
>       }
>    }
> 
>    return iterator(newNode.release());
> }
> 
> template <typename T, typename M, typename Alloc>
558d601
<    std::lock_guard<M> guard(m_write_mutex);
577d619
<    std::lock_guard<M> guard(m_write_mutex);
595d636
<    std::lock_guard<M> guard(m_write_mutex);
614d654
<    std::lock_guard<M> guard(m_write_mutex);
632,633d671
<    std::lock_guard<M> guard(m_write_mutex);
< 

```

**Comment by hongnod** [comment]
I did a search to commits, found following commit, but I don't if is it relative.
``


> > I did some search, it possibly relates to "try_lock" in "copperspice/src/annex/cs_libguarded/cs_shared_guarded.h" .
> 
> The CsLibGuarded library is not new and has been in CS for over six years. This particular header file has been present and unchanged in the last two releases of CopperSpice. We simply moved this file from CsCore to a new location.

I did a commits search， and found following one, but don't if is it relative
`SHA-1: 4b3cc715f450898dd0cbc18eb05ba1af1c6596d1

* Move cs_libguarded source from src/core/guarded to src/annex/cs_libguarded
10c10
< * CopperSpice is distributed in the hope that it will be useful,
---
> * CsLibGuarded is distributed in the hope that it will be useful,
233,234c233,234
<       void rcu_write_lock(const rcu_list<T, M, Alloc> &list);
<       void rcu_write_unlock(const rcu_list<T, M, Alloc> &list);
---
>       void rcu_write_lock(rcu_list<T, M, Alloc> &list);
>       void rcu_write_unlock(rcu_list<T, M, Alloc> &list);
306c306
< void rcu_list<T, M, Alloc>::rcu_guard::rcu_write_lock(const rcu_list<T, M, Alloc> &list)
---
> void rcu_list<T, M, Alloc>::rcu_guard::rcu_write_lock(rcu_list<T, M, Alloc> &list)
308a309
>    list.m_write_mutex.lock();
312c313
< void rcu_list<T, M, Alloc>::rcu_guard::rcu_write_unlock(const rcu_list<T, M, Alloc> &list)
---
> void rcu_list<T, M, Alloc>::rcu_guard::rcu_write_unlock(rcu_list<T, M, Alloc> &list)
313a315
>    list.m_write_mutex.unlock();
555a558,599
> template <typename... Us>
> auto rcu_list<T, M, Alloc>::emplace(const_iterator iter, Us &&...vs) -> iterator
> {
>    auto newNode = detail::allocate_unique<node>(m_node_alloc, std::forward<Us>(vs)...);
> 
>    node *oldHead = m_head.load();
>    node *oldTail = m_tail.load();
> 
>    if (oldHead == nullptr) {
>       // inserting into an empty list
>       m_head.store(newNode.get());
>       m_tail.store(newNode.get());
> 
>    } else if (oldHead == iter.m_current) {
>       // inserting at the beginning of a non-empty list
>       newNode->next.store(oldHead);
>       oldHead->back.store(newNode.get());
>       m_head.store(newNode.get());
> 
>    } else if (oldTail == iter.m_current) {
>       // inserting at the end of a non-empty list
>       newNode->back.store(oldTail);
>       oldTail->next.store(newNode.get());
>       m_tail.store(newNode.get());
> 
>    } else {
>       // inserting in the middle of a non-empty list
>       node *oldBack = iter.m_current->back.load();
> 
>       newNode->next.store(iter.m_current);
>       newNode->back.store(oldBack);
>       iter.m_current->back.store(newNode.get());
> 
>       if (oldBack != nullptr) {
>          oldBack->next.store(newNode.get());
>       }
>    }
> 
>    return iterator(newNode.release());
> }
> 
> template <typename T, typename M, typename Alloc>
558d601
<    std::lock_guard<M> guard(m_write_mutex);
577d619
<    std::lock_guard<M> guard(m_write_mutex);
595d636
<    std::lock_guard<M> guard(m_write_mutex);
614d654
<    std::lock_guard<M> guard(m_write_mutex);
632,633d671
<    std::lock_guard<M> guard(m_write_mutex);
< 
`

**Comment by hongnod** [comment]
> > I did some search, it possibly relates to "try_lock" in "copperspice/src/annex/cs_libguarded/cs_shared_guarded.h" .
> 
> The CsLibGuarded library is not new and has been in CS for over six years. This particular header file has been present and unchanged in the last two releases of CopperSpice. We simply moved this file from CsCore to a new location.

I did a search to commits, found following commit, but I don't if is it relative.
```
SHA-1: 4b3cc715f450898dd0cbc18eb05ba1af1c6596d1

* Move cs_libguarded source from src/core/guarded to src/annex/cs_libguarded
10c10
< * CopperSpice is distributed in the hope that it will be useful,
---
> * CsLibGuarded is distributed in the hope that it will be useful,
233,234c233,234
<       void rcu_write_lock(const rcu_list<T, M, Alloc> &list);
<       void rcu_write_unlock(const rcu_list<T, M, Alloc> &list);
---
>       void rcu_write_lock(rcu_list<T, M, Alloc> &list);
>       void rcu_write_unlock(rcu_list<T, M, Alloc> &list);
306c306
< void rcu_list<T, M, Alloc>::rcu_guard::rcu_write_lock(const rcu_list<T, M, Alloc> &list)
---
> void rcu_list<T, M, Alloc>::rcu_guard::rcu_write_lock(rcu_list<T, M, Alloc> &list)
308a309
>    list.m_write_mutex.lock();
312c313
< void rcu_list<T, M, Alloc>::rcu_guard::rcu_write_unlock(const rcu_list<T, M, Alloc> &list)
---
> void rcu_list<T, M, Alloc>::rcu_guard::rcu_write_unlock(rcu_list<T, M, Alloc> &list)
313a315
>    list.m_write_mutex.unlock();
555a558,599
> template <typename... Us>
> auto rcu_list<T, M, Alloc>::emplace(const_iterator iter, Us &&...vs) -> iterator
> {
>    auto newNode = detail::allocate_unique<node>(m_node_alloc, std::forward<Us>(vs)...);
> 
>    node *oldHead = m_head.load();
>    node *oldTail = m_tail.load();
> 
>    if (oldHead == nullptr) {
>       // inserting into an empty list
>       m_head.store(newNode.get());
>       m_tail.store(newNode.get());
> 
>    } else if (oldHead == iter.m_current) {
>       // inserting at the beginning of a non-empty list
>       newNode->next.store(oldHead);
>       oldHead->back.store(newNode.get());
>       m_head.store(newNode.get());
> 
>    } else if (oldTail == iter.m_current) {
>       // inserting at the end of a non-empty list
>       newNode->back.store(oldTail);
>       oldTail->next.store(newNode.get());
>       m_tail.store(newNode.get());
> 
>    } else {
>       // inserting in the middle of a non-empty list
>       node *oldBack = iter.m_current->back.load();
> 
>       newNode->next.store(iter.m_current);
>       newNode->back.store(oldBack);
>       iter.m_current->back.store(newNode.get());
> 
>       if (oldBack != nullptr) {
>          oldBack->next.store(newNode.get());
>       }
>    }
> 
>    return iterator(newNode.release());
> }
> 
> template <typename T, typename M, typename Alloc>
558d601
<    std::lock_guard<M> guard(m_write_mutex);
577d619
<    std::lock_guard<M> guard(m_write_mutex);
595d636
<    std::lock_guard<M> guard(m_write_mutex);
614d654
<    std::lock_guard<M> guard(m_write_mutex);
632,633d671
<    std::lock_guard<M> guard(m_write_mutex);
< 

```

**Comment by bgeller** [comment]
>  build program against cs-1.81 and run with cs-ver-1.80 dll

Please know this combination is not a valid test since CopperSpice does not guarantee ABI compatibility. You can not mix and match DLL files from different builds.

**Comment by bgeller** [comment]
The SHA you mentioned for CsLibGuarded added support for the method emplace() which was a request by another user. This method is not used in CsCore so it would have no effect on your application. 

We will look at your sample project and see if we can repeat this. Thanks for providing the zip file.

**Comment by bgeller** [comment]
Your issue was tested on several Unix platforms and it was not repeatable.  However, we were able to repeat the problem on MSVC in debug mode. Sadly, it is not uncommon for the MSVC compiler to show different results.  A CS team member as been assigned to investigate this problem.

**Comment by hongnod** [comment]
> 

There is an another change, as 
![image](https://user-images.githubusercontent.com/6680267/229385252-95d7a418-e554-42ae-846d-b60161f845e2.png)

**Comment by hongnod** [comment]
> The SHA you mentioned for CsLibGuarded added support for the method emplace() which was a request by another user. This method is not used in CsCore so it would have no effect on your application.
> 
> We will look at your sample project and see if we can repeat this. Thanks for providing the zip file.

thanks!

**Comment by bgeller** [comment]
Just so you know, we are very much aware of the changes to CsLibGuarded since this is a library we designed, developed and maintain. The git repository for that library is part of the CopperSpice project.  No need to let us know about the changes we made.

As I mentioned, there are no issues on any of the other platforms in release or debug mode. The only place this shows up in MSVC debug.

**Comment by agserm** [comment]
This issue has been resolved in SHA ca5a823.
